### DISCUSSION TITLE
Better Performance for concurrent column aggregation queries

### DISCUSSION THREAD

**Discussion by suimong** [duckdb/duckdb#2502] [Q&A] [answered]
Hi everyone, 

I have a dataset A that's about 10mm rows * 1000 columns, partitioned by year, and stored in 11 parquet files. There's another single parquet file B which also has a "date" column and a "rate" column. My query pattern looks like this: 
`select count(B.rate), sum(B.rate) from (A inner join B using date) groupby A.col0001`
where col0001 is one of 1000columns in A, and this query needs to be run on col0002, col0003, ... to col0999.

Currently I'm just doing a for loop on the columns in A, which took duckdb ~900ms per cycle (using conn.execute("sql query").fetchnumpy()) and ~15minutes to finish the loop. My question is, with the constraints of limited RAM (32G, which is not enough to read in all the data into memory and perform aggregation), is there a better query pattern for such task?

**Comment by Mytherin** [comment, marked as answer]
The latest version of DuckDB supports [Grouping Sets](https://duckdb.org/docs/sql/query_syntax/grouping_sets). Perhaps that could be of use here, e.g.:

```sql
SELECT count(B.rate),
       sum(B.rate)
FROM (A INNER JOIN B USING date)
GROUP BY 
GROUPING SETS ((A.col0001), (A.col0002), ...)
```

**Reply by suimong** [reply]
Hi @Mytherin , thanks for the super quick and useful tip! Just realized that `GROUPING SETS` is not in the 0.3.0 release, so it'll take me a while to build duckdb from source. I'll report back on my findings. Thanks for making such a wonderful package, really appreciate and admire your work!

**Reply by hannes** [reply]
Next release should come within the next week or so too
